hazelcast/src/hazelcast/util/impl/AbstractThread.cpp

```diff
@@ -24,11 +24,24 @@ namespace hazelcast {
         namespace impl {
             util::SynchronizedMap<int64_t, AbstractThread::UnmanagedAbstractThreadPointer> AbstractThread::startedThreads;
 
-            AbstractThread::AbstractThread(const boost::shared_ptr<Runnable> &runnable) : target(runnable) {
+            /**
+             * Latch count is 2, because one is count down when thread started and the second one is decremented when
+             * thread exits. Derived classes should decrement this latch just before exiting the thread
+             * execution method.
+             * @param runnable The runnable to run when this thread is started.
+             */
+            AbstractThread::AbstractThread(const boost::shared_ptr<Runnable> &runnable) : target(runnable), stateLatch(
+                    new util::CountDownLatch(2)) {
                 this->target = runnable;
             }
 
             AbstractThread::~AbstractThread() {
+                if (started) {
+                    int state = stateLatch->get();
+                    if (state == 1) {
```

Comment:
Why is this `if` statement necessary.  It looks like the second count is doing same thing as `started` boolean. I would remove the second count that tracks started. 

And there seems to be still race between destructor and start method. 
When both starts at the same time, it could be the case that destructor passes then thread starts. 
To prevent that, ( I am asking because I am not sure about usages of abstractThread )
there needs to be a third state and `started`  that we set on destructor so that `start` will be prevented.

Reply:
Changed the Thread implementation to use state machine. The states are :
                enum ThreadState {
                    UNSTARTED = 0,
                    STARTED = 1,
                    JOINED = 2,
                    CANCELLED = 3
                };
It can go from UNSTARTED to STARTED, STARTED to JOINED or STARTED to CANCELLED states. Other transitions are not valid. When thread is cancelled, then the shutdown latch is waited.